registry/session_v2.go

```diff
@@ -19,10 +20,27 @@ func getV2Builder(e *Endpoint) *v2.URLBuilder {
 	return e.URLBuilder
 }
 
-func (r *Session) V2RegistryEndpoint(index *IndexInfo) (ep *Endpoint, err error) {
-	// TODO check if should use Mirror
-	if index.Official {
-		ep, err = newEndpoint(REGISTRYSERVER, true)
+func (r *Session) V2RegistryEndpoint(index *IndexInfo, useMirror bool) (ep *Endpoint, err error) {
+	if useMirror && len(index.Mirrors) > 1 {
+		for i := 0; i < len(index.Mirrors) && ep == nil; i++ {
+			ep, err = newEndpoint(index.Mirrors[0], index.Secure)
+			if err != nil {
+				return
+			}
+			err = validateEndpoint(ep)
+			if err != nil {
+				return
+			}
+			if ep.Version == APIVersion1 {
+				ep = nil
```

Comment:
why this ep==nil could we not return "no V@ supported mirrors" here, also why `&& ep ==nil` on line 25, are you trying to stop the loop when you have found one?, why not use `break`